README.rst
disqus-python
~~~~~~~~~~~~~

Let's start with installing the API:

.. code:: shell

    $ pip install disqus-python

Use the API by instantiating it, and then calling the method through dotted notation chaining:

.. code:: python

    from disqusapi import DisqusAPI
    disqus = DisqusAPI(secret_key, public_key)
    for result in disqus.trends.listThreads():
        print result

Parameters (including the ability to override version, api_secret, and format) are passed as 
keyword arguments to the resource call:

.. code:: python

    disqus.posts.details(post=1, version='3.0')

Paginating through endpoints is easy as well:

.. code:: python

    from disqusapi import Paginator
    paginator = Paginator(api.trends.listThreads, forum='disqus')
    for result in paginator:
        print result

    # pull in a maximum of 500 results (this limit param differs from the endpoint's limit param)
    for result in paginator(limit=500):
        print result

Documentation on all methods, as well as general API usage can be found at http://disqus.com/api/


CHANGES
NEXT VERSION

* Improved README.rst

0.4.1

* Fixed behavior when api_secret or api_public are not set.

0.4.0

* Removed signed requests (Disqus has deprecated them).
* Change all endpoints to use SSL.

0.3.2

* Added support for undefined interfaces.
* Added support for mapping error codes to different exceptions.

0.3.1

* Fixed an issue with GET requests and the normalized request string.

0.3.0

* Added signed requests (you must pass public_key to DisqusAPI).
* Added support for OAuth (via access_token in signed requests).
* Moved Paginator into disqusapi.paginator


disqusapi/tests/__init__.py
import sys


if sys.version_info >= (2, 7):
    import unittest
else:
    import unittest2 as unittest


disqusapi/tests/test_api.py
from mock import Mock

from disqusapi import (
    Result,
    ResourceElement,
    Resource,
    DisqusRequest,
    params_list,
    DisqusAPI,
    load_interfaces)
from disqusapi.exceptions import (
    InterfaceNotDefined,
    APIError,
    InvalidAccessToken,
    RateLimitError)
from disqusapi.tests import unittest


class TestResult(unittest.TestCase):
    def test_init_response(self):
        self.assertEqual('hey', Result('hey').response)

    def test_init_cursor_given(self):
        self.assertEqual('cursor', Result('hey', cursor='cursor').cursor)

    def test_init_cursor_none(self):
        self.assertEqual({}, Result('hey').cursor)

    def test_repr(self):
        self.assertEqual('<Result: 1234>', repr(Result(1234)))

    def test_iter(self):
        self.assertEqual([0, 1], list(Result(range(2))))

    def test_len(self):
        self.assertEqual(6, len(Result('banana')))

    def test_slice(self):
        """Slice works without the deprecated getslice"""
        self.assertEqual([1, 3], Result(list(range(15)))[1:5:2])

    def test_getitem(self):
        self.assertEqual(1, Result([0, 1, 2])[1])

    def test_contains(self):
        self.assertTrue('a' in Result('banana'))


class MyResourceElement(ResourceElement):
    def _new_element(self, interface, attr, tree):
        return MyResourceElement(interface, attr, tree)


class TestResourceElement(unittest.TestCase):
    def test_init_interface(self):
        self.assertEqual(
            'interfaces',
            MyResourceElement('interfaces', None, ()).interface)

    def test_init_tree(self):
        self.assertEqual((), MyResourceElement(1, None, ()).tree)

    def test_init_node_changes_tree(self):
        self.assertEqual(
            (1, 2),
            MyResourceElement('iface', 2, (1,)).tree)

    def test_eq(self):
        one = MyResourceElement('iface', 2, (1,))
        two = MyResourceElement('iface', 2, (1,))
        self.assertEqual(one, two)

    def test_not_eq_iface(self):
        one = MyResourceElement('iface', node=2, tree=(1,))
        two = MyResourceElement('', node=2, tree=(1,))
        self.assertNotEqual(one, two)

    def test_getattr(self):
        interface = {'apple': {'orange': {'wine': 'good'}}}
        sut = MyResourceElement(interface, None, ()).apple.orange.wine
        expected = MyResourceElement('good', 'wine', ('apple', 'orange'))
        self.assertEqual(expected, sut)

    def test_getattr_undefined_interface(self):
        sut = MyResourceElement({}, 2, ()).supercalifragilisticexpialidocious
        self.assertEqual({}, sut.interface)

    def test_new_element(self):
        with self.assertRaises(NotImplementedError):
            ResourceElement({}, None, ()).test


class TestResource(unittest.TestCase):
    def test_init_request(self):
        self.assertEqual('req', Resource('req', {}, None, ()).request)

    def test_resource_element_api(self):
        sut = Resource('req', {}, None, ())
        self.assertEqual('req', sut.hey.request)

    def test_invalid_argument(self):
        with self.assertRaises(ValueError):
            Resource('req', {'required': ['test']}, None, ())()

    def test_invalid_method(self):
        with self.assertRaises(InterfaceNotDefined):
            Resource('req', {}, None, ())()

    def test_request(self):
        request = Mock()
        Resource(request, {'method': 'POST'}, 'b', ('a',))(c=1)
        request.assert_called_with('POST', 'a/b', {'c': 1})

    def test_request_required_query_parameter(self):
        request = Mock()
        interface = {'method': 'POST', 'required': ['c']}
        Resource(request, {'method': 'POST'}, 'b', ('a',))(**{'c:ident': 1})
        request.assert_called_with('POST', 'a/b', {'c:ident': 1})


class TestDisqusRequest(unittest.TestCase):
    def setUp(self):
        self.conn = Mock()
        self.request = self.conn.return_value.request
        response = self.conn.return_value.getresponse.return_value
        response.status = 200
        response.read.return_value = '{"response": 1}'
        self.response = response
        self.sut = DisqusRequest('secret', '3.0', self.conn)

    def set_error_code(self, code):
        self.response.status = 500
        self.response.read.return_value = (
            '{"response": [1, 2], "code": %d}' % code)

    def test_update_keys_secret(self):
        kwargs = {}
        self.sut('POST', 'a/b', kwargs)
        self.assertEqual('secret', kwargs['api_secret'])

    def test_connection(self):
        self.sut('POST', 'a/b', {})
        self.conn.assert_called_with(self.sut.host)

    def test_request_call(self):
        self.sut('POST', 'a/b', {})
        self.request.assert_called_with(
            'POST', '/api/3.0/a/b.json', 'api_secret=secret', self.sut.headers)

    def test_request_get(self):
        self.sut('GET', 'a/b', {})
        self.request.assert_called_with(
            'GET', '/api/3.0/a/b.json?api_secret=secret', '', self.sut.headers)

    def test_response(self):
        self.assertEqual(1, self.sut('POST', 'a/b', {}))

    def test_response_list(self):
        self.response.read.return_value = '{"response": [1, 2], "cursor": 3}'
        res = self.sut('POST', 'a/b', {})
        self.assertEqual(Result([1, 2], 3), res)

    def test_error(self):
        self.set_error_code(1)
        with self.assertRaises(APIError):
            self.sut('POST', 'a/b', {})

    def test_error_13(self):
        self.set_error_code(13)
        with self.assertRaises(RateLimitError):
            self.sut('POST', 'a/b', {})

    def test_error_14(self):
        self.set_error_code(14)
        with self.assertRaises(RateLimitError):
            self.sut('POST', 'a/b', {})

    def test_error_18(self):
        self.set_error_code(18)
        with self.assertRaises(InvalidAccessToken):
            self.sut('POST', 'a/b', {})


class TestParamsList(unittest.TestCase):
    def test_simple(self):
        self.assertEqual([('a', 1)], params_list({'a': 1}))

    def test_list(self):
        self.assertEqual([('a', 1), ('a', 2)], params_list({'a': [1, 2]}))

    def test_tuple(self):
        self.assertEqual([('a', 1), ('a', 2)], params_list({'a': (1, 2)}))


class TestDisqusAPI(unittest.TestCase):
    def test_init_interface(self):
        self.assertEqual(load_interfaces(), DisqusAPI('secret').interface)

    def test_init_tree(self):
        self.assertEqual((), DisqusAPI('secret').tree)

    def test_resource_element(self):
        sut = DisqusAPI('secret')
        self.assertEqual(
            Resource(sut.make_request, {}, 'hey', ()),
            sut.hey)


disqusapi/tests/test_paginator.py
from disqusapi.tests import unittest
from disqusapi import Result
from disqusapi.exceptions import RateLimitError
from disqusapi.paginator import (
    ignore_limit,
    use_cursor,
    limit_amount,
    Paginator)


class TestIgnoreLimit(unittest.TestCase):
    def test_ok(self):
        def endpoint():
            return 1
        self.assertEqual(1, ignore_limit(endpoint)())

    def test_rate_limit(self):
        def endpoint():
            raise RateLimitError(14, 'error')
        self.assertEqual(Result([]), ignore_limit(endpoint)())

    def test_other_error(self):
        def endpoint():
            raise ValueError('error')
        with self.assertRaises(ValueError):
            ignore_limit(endpoint)()


class TestUseCursor(unittest.TestCase):
    def test_no_pages(self):
        def endpoint():
            return Result([1, 2])
        self.assertEqual([1, 2], list(use_cursor(endpoint)()))

    def test_one_page(self):
        def endpoint(cursor=None):
            if cursor is None:
                return Result([1, 2], cursor=dict(more=True, id=3))
            else:
                return Result([cursor])
        self.assertEqual([1, 2, 3], list(use_cursor(endpoint)()))


class TestLimitAmount(unittest.TestCase):
    def test_limited(self):
        def endpoint():
            return range(100)
        self.assertEqual(
            list(range(10)),
            list(limit_amount(endpoint, 10)()))

    def test_limit_too_big(self):
        def endpoint():
            return range(2)

        self.assertEqual([0, 1], list(limit_amount(endpoint, 10)()))


class TestPaginator(unittest.TestCase):
    def test_simple(self):
        def endpoint():
            return Result([1, 2])
        self.assertEqual([1, 2], list(Paginator(endpoint)))

    def test_limit(self):
        def endpoint():
            return Result([1, 2, 3, 4])
        self.assertEqual([1, 2], list(Paginator(endpoint)(limit=2)))

    def test_silence_limit(self):
        def endpoint(cursor=None):
            if cursor is not None:
                raise RateLimitError(14, 'error')
            else:
                return Result([1, 2], dict(more=True, id=3))
        self.assertEqual([1, 2], list(Paginator(endpoint)(silence_limit=True)))


disqusapi/utils.py
import binascii
import hashlib
import hmac

from disqusapi.compat import urlencode, urlparse


def get_normalized_params(params):
    """
    Given a list of (k, v) parameters, returns
    a sorted, encoded normalized param
    """
    return urlencode(sorted(params))


def get_normalized_request_string(
        method, url, nonce, params, ext='', body_hash=None):
    """
    Returns a normalized request string as described iN OAuth2 MAC spec.

    http://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-00#section-3.3.1
    """
    urlparts = urlparse(url)
    if urlparts.query:
        norm_url = '%s?%s' % (urlparts.path, urlparts.query)
    elif params:
        norm_url = '%s?%s' % (urlparts.path, get_normalized_params(params))
    else:
        norm_url = urlparts.path

    if not body_hash:
        body_hash = get_body_hash(params)

    port = urlparts.port
    if not port:
        assert urlparts.scheme in ('http', 'https')

        if urlparts.scheme == 'http':
            port = 80
        elif urlparts.scheme == 'https':
            port = 443

    output = [nonce, method.upper(), norm_url, urlparts.hostname, port,
              body_hash, ext, '']

    return '\n'.join(map(str, output))


def get_body_hash(params):
    """
    Returns BASE64 ( HASH (text) ) as described in OAuth2 MAC spec.

    http://tools.ietf.org/html/draft-ietf-oauth-v2-http-mac-00#section-3.2
    """
    norm_params = get_normalized_params(params)

    return binascii.b2a_base64(hashlib.sha1(norm_params).digest())[:-1]


def get_mac_signature(api_secret, norm_request_string):
    """
    Returns HMAC-SHA1 (api secret, normalized request string)
    """
    hashed = hmac.new(str(api_secret), norm_request_string, hashlib.sha1)
    return binascii.b2a_base64(hashed.digest())[:-1]


disqusapi/paginator.py
"""Cursor and other goodies for DisquisAPI methods"""
from disqusapi import Result
from disqusapi.exceptions import RateLimitError


class Paginator(object):
    """
    Paginate through all entries:

    >>> from disqusapi.paginator import Paginator
    >>> from disqusapi import DisqusAPI
    >>> api = DisqusAPI('secret_key')
    >>> paginator = Paginator(api.trends.listThreads, forum='disqus')
    >>> for result in paginator:
    >>>     print(result)

    Paginate only up to a number of entries:

    >>> for result in paginator(limit=500):
    >>>     print(result)

    Paginate till you run out of api calls:

    >>> for result in paginator(silence_limit=True):
    >>>     print(result)
    """

    def __init__(self, endpoint, **params):
        self.endpoint = endpoint
        self.params = params

    def __iter__(self):
        for result in self():
            yield result

    def __call__(self, limit=None, silence_limit=False):
        endpoint = self.endpoint
        if silence_limit:
            endpoint = ignore_limit(endpoint)
        endpoint = use_cursor(endpoint)
        if limit is not None:
            endpoint = limit_amount(endpoint, limit)
        for result in endpoint(**self.params):
            yield result


def limit_amount(endpoint, limit):
    def wrapped(**kwargs):
        count = 0
        for result in endpoint(**kwargs):
            yield result
            count += 1
            if limit == count:
                break
    return wrapped


def use_cursor(endpoint):
    def wrapped(**kwargs):
        while True:
            results = endpoint(**kwargs)
            for result in results:
                yield result

            if results.cursor and results.cursor['more']:
                kwargs['cursor'] = results.cursor['id']
            else:
                break
    return wrapped


def ignore_limit(endpoint):
    def wrapped(**kwargs):
        try:
            return endpoint(**kwargs)
        except RateLimitError:
            return Result([])
    return wrapped


disqusapi/__init__.py
"""
disqus-python
~~~~~~~~~~~~~

disqus = DisqusAPI(secret_key)
disqus.trends.listThreads()

"""
try:
    __version__ = __import__('pkg_resources') \
        .get_distribution('disqusapi').version
except:
    __version__ = 'unknown'

import os.path
import simplejson

from disqusapi.compat import urlencode, HTTPSConnection
from disqusapi.exceptions import (
    InterfaceNotDefined,
    APIError,
    InvalidAccessToken,
    RateLimitError)

__all__ = ['DisqusAPI']


class Result(object):
    """
    Wraps a result dictionary and its cursor.

    It allows common operations directly over the response part.
    """
    def __init__(self, response, cursor=None):
        self.response = response
        self.cursor = cursor or {}

    def __repr__(self):
        return '<Result: %s>' % repr(self.response)

    def __iter__(self):
        for result in self.response:
            yield result

    def __len__(self):
        return len(self.response)

    def __getitem__(self, key):
        return self.response.__getitem__(key)

    def __contains__(self, key):
        return self.response.__contains__(key)

    def __eq__(self, other):
        return self.response == other.response and \
            self.cursor == other.cursor


class ResourceElement(object):
    """
    Object like walking resource generation

    This allows resource generation in this form:

    >>> this.is.an.api.resource.now

    Generating resources for each step, so in the end you get a resource with a
    tree like this: `('this', 'is', 'an', 'api', 'resource')`

    Since at least two classes use this functionality, do not forget to
    override _new_element. This method specify what to build as a resource and
    how to build it.
    """
    def __init__(self, interface, node, tree):
        self.interface = interface
        if node:
            tree = tree + (node,)
        self.tree = tree

    def __eq__(self, other):
        return self.interface == other.interface and \
            self.tree == other.tree

    def __getattr__(self, attr):
        interface = self.interface
        if attr not in interface:
            interface[attr] = {}
        return self._new_element(interface[attr], attr, self.tree)

    def _new_element(self, interface, node, tree):
        """What to build as a resource in the next step (and how to do it)"""
        raise NotImplementedError('Implment in your subclass')


class Resource(ResourceElement):
    def __init__(self, request, interface, node, tree):
        super(Resource, self).__init__(interface, node, tree)
        self.request = request

    def _new_element(self, interface, node, tree):
        return Resource(self.request, interface, node, tree)

    def __eq__(self, other):
        return self.request == other.request and \
            super(Resource, self).__eq__(other)

    def __call__(self, **kwargs):
        return self._make_request(**kwargs)

    def _validate_arguments(self, kwargs):
        """Validate arguments taking care of queries in the names"""
        keys = set(key.split(':')[0] for key in kwargs.keys())
        for required in self.interface.get('required', []):
            if required not in keys:
                raise ValueError('Missing required argument: %s' % required)

    def _validate_method(self, kwargs):
        method = kwargs.pop('method', self.interface.get('method'))
        if not method:
            raise InterfaceNotDefined(
                'Interface is not defined, you must pass ``method``'
                ' (HTTP Method).')
        return method

    def _make_request(self, **kwargs):
        self._validate_arguments(kwargs)
        method = self._validate_method(kwargs)
        return self.request(method, '/'.join(self.tree), kwargs)


class DisqusRequest(object):
    headers = {'User-Agent': 'disqus-python/%s' % __version__}
    error_map = {
        13: RateLimitError,
        14: RateLimitError,
        18: InvalidAccessToken,
    }
    host = 'disqus.com'

    def __init__(self, secret_key, version, conn=HTTPSConnection):
        self.__secret_key = secret_key
        self.__version = version
        self.__conn = conn

    def _update_keys(self, kwargs):
        if 'api_secret' not in kwargs and self.__secret_key:
            kwargs['api_secret'] = self.__secret_key

    def __call__(self, method, path, kwargs):
        self._update_keys(kwargs)
        params = params_list(kwargs)
        # Adjust path
        path = '/api/%s/%s.json' % (self.__version, path)

        # Adjust data based on the method
        if method == 'GET':
            path = '%s?%s' % (path, urlencode(params))
            data = ''
        else:
            data = urlencode(params)

        conn = self.__conn(self.host)
        conn.request(method, path, data, self.headers)

        response = conn.getresponse()
        # Let's coerce it to Python
        data = simplejson.loads(response.read())

        if response.status != 200:
            exception_class = self.error_map.get(data['code'], APIError)
            raise exception_class(data['code'], data['response'])

        if isinstance(data['response'], list):
            return Result(data['response'], data.get('cursor'))
        return data['response']


def params_list(kwargs):
    params = []
    for key, value in kwargs.items():
        if isinstance(value, (list, tuple)):
            params.extend([(key, list_val) for list_val in value])
        else:
            params.append((key, value))
    return params


def load_interfaces():
    return simplejson.loads(open(os.path.join(
        os.path.dirname(__file__), 'interfaces.json'), 'r').read())


class DisqusAPI(ResourceElement):
    def __init__(self, secret_key=None, version='3.0', **kwargs):
        super(DisqusAPI, self).__init__(load_interfaces(), None, ())
        self.make_request = DisqusRequest(secret_key, version)

    def _new_element(self, interface, node, tree):
        return Resource(self.make_request, interface, node, tree)
